### DISCUSSION TITLE
Unexpected behavior with Composite Functions

### DISCUSSION THREAD

**Discussion by jaylevin** [crossplane/crossplane#3905] [Q&A]
I have been experimenting with composite functions for the past week and have some questions regarding the intended behavior of them.

I have a composition function very similar to the example in the documentation. It adds an annotation to a specific resource in the `functionio["desired"]["resources"]` object. I have a conditional case to check if the annotation already exists in the desired resource, and if it doesn't it adds it. If the annotation already exists, I also add a message to the results array to help with debugging.

The function looks something like this:
```python

    # basic function to add an annotation if it does not exist
    try:
        for r in functionIO["desired"]["resources"]:
            result_normal(functionIO, f"resource: {r}")
            if r["name"] == "environment":
                if "resource" not in r:
                    continue
                if "metadata" not in r["resource"]:
                    r["resource"]["metadata"] = {}
                if "annotations" not in r["resource"]["metadata"]:
                    r["resource"]["metadata"]["annotations"] = {}

                ANNOTATION_NAME = "test-annotation"
                if ANNOTATION_NAME not in r["resource"]["metadata"]["annotations"]:
                    # should only run once per resource
                    result_normal(functionIO, "desired resource does not have annotation, one will be added!")
                    r["resource"]["metadata"]["annotations"][ANNOTATION_NAME] = "test"
                else:
                    result_normal(functionIO, "desired resource has annotation, skipping!")
                break
    except Exception as e:
        result_normal(functionIO, f"exception: {str(e)}")
```

The problem I am running into is the desired object does not seem to be saved between function executions.

For example:
1. The function runs the first time and adds the annotation to the desired object and returns.
2. Reconcile happens and the annotation is added to the composed resource.
3. The function runs a second time, and once again, the desired resource in the `FunctionIO` object does not have the annotation. I would expect that the desired object being passed to the function would contain the changes that were made in the previous function execution. Is this not the intended behavior, or is this a bug? Should I be checking the observed resource instead of the desired resource?

I have tried using Crossplane v1.11.1, v1.11.2, and the latest build on the master branch ([crossplane-1.12.0-rc.0.213.g2a65adbb.tgz](https://charts.crossplane.io/master/crossplane-1.12.0-rc.0.213.g2a65adbb.tgz))

**Comment by jbw976** [comment]
This has been opened as an issue in #3918 and is being discussed there 😄
